crates/typst-library/src/layout/grid/resolve.rs

```diff
@@ -1615,25 +1608,190 @@ impl<'a> CellGrid<'a> {
+    //
+    // y = 0 => occupied
+    // y = 1 => empty (first_available_row = 1)
+    // y = 2 => occupied
+    // y = 3 => header
+    //
+    // Here, we should bail if the header tries to expand upwards, regardless
+    // of the fact that the conflicting row (y = 2) comes after the first
+    // available row.
+    //
+    // Note that expanding upwards is only possible when row-positioned cells
+    // are specified, in one of the following cases:
+    //
+    // 1. We place e.g. 'table.cell(y: 3)' followed by 'table.cell(y: 2)'
+    // (earlier row => upwards);
+    //
+    // 2. We place e.g. 'table.cell(y: 3)' followed by '[a]' (auto-pos cell
+    // favors 'first_available_row', so the header tries to expand upwards to
+    // place the cell at 'y = 1' and conflicts at 'y = 2') or
+    // 'table.cell(x: 1)' (same deal).
+    //
+    // Of course, we also need to check for downward expansion as usual as
+    // there could be a non-empty row below the header, but the upward case is
+    // highlighted as it was checked separately before (and also to explain
+    // what kind of situation we are preventing with this check).
+    //
+    // Note that simply checking for non-empty rows like below not only
+    // prevents conflicts with top-level cells (outside of headers and
+    // footers), but also prevents conflicts with other headers or footers,
+    // since we have an invariant that even empty headers and footers must
+    // contain at least one 'Some(...)' position in 'resolved_cells'. More
+    // precisely, each header and footer has at least one 'Some(...)' cell at
+    // 'group_range.start' and at 'group_range.end - 1' - non-empty headers and
+    // footers don't span any unnecessary rows. Therefore, we don't have to
+    // loop over headers and footers, only check if the new rows are empty.
+    for new_y in new_rows {
+        if let Some(new_row @ [_non_empty, ..]) = resolved_cells
+            .get(new_y * columns..)
+            .map(|cells| &cells[..columns.min(cells.len())])
+        {
+            if new_row.iter().any(Option::is_some) {
+                // TODO:
+                // - Later/earlier rows might be confusing
+                // (moving to the end always works...)
+                // - Detect when header or footer collided with
+                // another header or footer and provide a
+                // better error message if so.
+                if group_range.is_none_or(|r| new_y < r.start) {
+                    bail!(
+                        "cell would cause header or footer to expand to non-empty row {new_y}";
+                        hint: "try moving its cells to later rows"
+                    );
+                } else {
+                    bail!(
+                        "cell would cause header or footer to expand to non-empty row {new_y}";
+                        hint: "try moving its cells to earlier rows"
+                    );
+                }
+            }
```

Comment:
A bit unsure about this error hint, thinking of just generically asking to move the header or footer, or the existing cells

Reply:
Changed it a bit, feel free to unresolve if you have further suggestions.